flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java

```diff
@@ -359,24 +425,45 @@ private void tryFulFillRedundantResources(
                         .map(internalResourceInfo -> internalResourceInfo.availableProfile)
                         .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
 
-        tryFulFillRedundantResourcesWithAction(
-                requiredRedundantResource,
+        ResourceProfile totalResources =
+                Stream.concat(
+                                availableRegisteredResources.stream(),
+                                availablePendingResources.stream())
+                        .map(internalResourceInfo -> internalResourceInfo.totalProfile)
+                        .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
+
+        tryFulFillRequiredResourcesWithAction(
                 totalAvailableResources,
+                totalResources,
                 resultBuilder::addPendingTaskManagerAllocate);
     }
 
-    private void tryFulFillRedundantResourcesWithAction(
-            ResourceProfile requiredRedundantResource,
-            ResourceProfile totalAvailableResources,
+    private void tryFulFillRequiredResourcesWithAction(
+            ResourceProfile resourceAvailable,
+            ResourceProfile resourceInTotal,
             Consumer<? super PendingTaskManager> fulfillAction) {
-        while (!canFulfillRequirement(requiredRedundantResource, totalAvailableResources)) {
+        while (!isRequiredResourcesFulfilled(resourceAvailable, resourceInTotal)) {
             PendingTaskManager pendingTaskManager =
                     new PendingTaskManager(totalResourceProfile, numSlotsPerWorker);
             fulfillAction.accept(pendingTaskManager);
-            totalAvailableResources = totalAvailableResources.merge(totalResourceProfile);
+            resourceAvailable = resourceAvailable.merge(totalResourceProfile);
+            resourceInTotal = resourceInTotal.merge(totalResourceProfile);
         }
     }
 
+    private ResourceProfile getTotalResourceOfTaskManagers(List<TaskManagerInfo> taskManagers) {
+        return taskManagers.stream()
+                .map(TaskManagerInfo::getTotalResource)
+                .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
```

Comment:
In this strategy, all the TM resource spec are the same. we can just use it to multiply the TM num for efficiency.